### PR TITLE
Upgrade Groovy to 4

### PR DESCRIPTION
**Pull request by donat**
*NOTE:* Superseded by https://github.com/gradle/gradle/pull/20038

Let's try out the new Groovy!

Problems found in Groovy `4.0.0-beta-1`:

- Dynamic Groovy code in Gradle doesn't compile because of $getLookup() method is not static - https://issues.apache.org/jira/browse/GROOVY-10290
- Groovy compiler generates invalid Java stubs  – https://issues.apache.org/jira/browse/GROOVY-10299

## TODO

- [ ] use new `AntBuilder` in our own – https://github.com/gradle/gradle/pull/18700
- [ ] upgrade Groovy version in `AbstractSourcesAndJavadocJarsIntegrationTest`
- [ ] use `org.apache.groovy` group ID for Groovy versions > 4 in `GroovyParametersMetadataIntegrationTest` (and probably in many other places)
- [ ] fix instrumentation to handle `INVOKEDYNAMIC`
- [ ] auto-import `groovy.xml` and `groovy.ant`? There are pros and cons...
- [ ] fix tests
- [ ] upgrade Spock once it's out – https://github.com/spockframework/spock/pull/1382
  - [ ] upgrade to Spock 2.2-M1
  - [x] remove all references to `spock.iKnowWhatImDoing.disableGroovyVersionCheck`
- [ ] upgrade to Groovy 4.0 GA
  - [ ] [GROOVY-10466](https://issues.apache.org/jira/browse/GROOVY-10466) has no plan to be fixed yet
  - [ ] [GROOVY-10467](https://issues.apache.org/jira/browse/GROOVY-10467) will be fixed in `4.0.1`
  - [x] [GROOVY-10299](https://issues.apache.org/jira/browse/GROOVY-10299) fixed in `4.0.1-beta-2`
  - [ ] [GROOVY-10290](https://issues.apache.org/jira/browse/GROOVY-10290) will be fixed in `4.0.1`
- [x] do what we can already on the 7.x line to prepare for the upgrade: https://github.com/gradle/gradle/pull/18665

### CODE REVIEW

**Comment by lptr** [comment]
We get a failure in `:internal-integ-test:compileGroovy`:

```text


:internal-integ-testing:compileGroovy FAILED |  
-- | --
  | Groovy compilation avoidance is an incubating feature. |  
  | Executing Spock 2.0.0-groovy-3.0 with NOT compatible Groovy version 4.0.0-beta-1 due to set spock.iKnowWhatImDoing.disableGroovyVersionCheck system property. This is unsupported and may result in weird runtime errors! |  
  | /Users/lptr/Workspace/gradle/gradle/subprojects/internal-integ-testing/build/tmp/compileGroovy/groovy-java-stubs/org/gradle/test/fixtures/server/sftp/SFTPServer.java:172: error: $getLookup() in SFTPServer.SftpExpectOneOpen cannot override $getLookup() in SftpExpectOnePath |  
  | public  java.lang.invoke.MethodHandles.Lookup $getLookup() { return (java.lang.invoke.MethodHandles.Lookup)null;} |  
  | ^ |  
  | overridden method is static |  
  | /Users/lptr/Workspace/gradle/gradle/subprojects/internal-integ-testing/build/tmp/compileGroovy/groovy-java-stubs/org/gradle/test/fixtures/server/sftp/SFTPServer.java:188: error: $getLookup() in SFTPServer.SftpExpectOneHandle cannot override $getLookup() in SftpExpectOnePath |  
  | public  java.lang.invoke.MethodHandles.Lookup $getLookup() { return (java.lang.invoke.MethodHandles.Lookup)null;} |  
  | ^ |  
  | overridden method is static |  
  | Note: Some input files use or override a deprecated API. |  
  | Note: Recompile with -Xlint:deprecation for details. |  
  | Note: Some input files use unchecked or unsafe operations. |  
  | Note: Recompile with -Xlint:unchecked for details. |  
  | 2 errors |  
  | startup failed: |  
  | Compilation failed; see the compiler error output for details. |  
  |   |  
  | 1 error


```

https://ge.gradle.org/s/nue34m3ucywh6/console-log?task=:internal-integ-testing:compileGroovy

**Comment by lptr** [comment]
I've filed a bug about the compile problem with Groovy: https://issues.apache.org/jira/browse/GROOVY-10290

I've added a tag to reproduce the problem: https://github.com/gradle/gradle/releases/tag/groovy-4-compile-problem

**Comment by lptr** [comment]
@bot-gradle test QFL please

**Comment by lptr** [comment]
@bot-gradle test QFL please.

**Comment by lptr** [comment]
@bot-gradle test QFL please.

**Comment by lptr** [comment]
@bot-gradle test this please

File: subprojects/testing-native/src/integTest/groovy/org/gradle/nativeplatform/test/cunit/CUnitTestResults.groovy [hunk 1]
```diff
@@ -27,7 +27,7 @@ class CUnitTestResults {
     CUnitTestResults(TestFile testResultsFile) {
         assert testResultsFile.exists()
         this.testResultsFile = testResultsFile
-        final XmlParser parser = new XmlParser(false, false)
+        final def parser = new XmlParser(false, false)
```
Comment (szpak) [inline subprojects/testing-native/src/integTest/groovy/org/gradle/nativeplatform/test/cunit/CUnitTestResults.groovy:30]: You could simplified it further to:
```suggestion
        final parser = new XmlParser(false, false)
```
(twice)

File: build-logic/build-platform/build.gradle.kts [hunk 2]
```diff
@@ -6,7 +6,7 @@ description = "Provides a platform that constrains versions of external dependen
 
 // Here you should declare versions which should be shared by the different modules of buildSrc itself
 val javaParserVersion = "3.18.0"
-val groovyVersion = "3.0.9"
+val groovyVersion = "4.0.0-beta-1"
```
Comment (szpak) [inline build-logic/build-platform/build.gradle.kts:9]: Why not 4.0.0?

**Comment by lptr** [comment]
@bot-gradle test this please.

File: build-logic/build-platform/build.gradle.kts (hunk 2 above)
Comment (lptr) [reply build-logic/build-platform/build.gradle.kts:9]: No, there are two bugs that prevent us from upgrading even to 4.0.0-beta-2:

https://issues.apache.org/jira/browse/GROOVY-10466
https://issues.apache.org/jira/browse/GROOVY-10467

File: subprojects/testing-native/src/integTest/groovy/org/gradle/nativeplatform/test/cunit/CUnitTestResults.groovy (hunk 1 above)
Comment (lptr) [reply subprojects/testing-native/src/integTest/groovy/org/gradle/nativeplatform/test/cunit/CUnitTestResults.groovy:30]: Yeah, there's tons of places where we could improve the code, but this PR is just about trying to pass our tests with Groovy 4.

**Comment by lptr** [comment]
@bot-gradle test this please

File: subprojects/testing-native/src/integTest/groovy/org/gradle/nativeplatform/test/cunit/CUnitTestResults.groovy (hunk 1 above)
Comment (szpak) [reply subprojects/testing-native/src/integTest/groovy/org/gradle/nativeplatform/test/cunit/CUnitTestResults.groovy:30]: Sure, I've just mentioned that one as you changed `XmlParser → def` (and could be `XmlParser → final`), with the same effort.

Comment (szpak) [reply subprojects/testing-native/src/integTest/groovy/org/gradle/nativeplatform/test/cunit/CUnitTestResults.groovy:30]: Nevertheless, it's just a nitpicking :).

**Comment by lptr** [comment]
@bot-gradle test this please

**Comment by lptr** [comment]
@bot-gradle test this please

**Comment by lptr** [comment]
@bot-gradle test this please

**Comment by lptr** [comment]
Superseded by https://github.com/gradle/gradle/pull/20038.

**Pull request closed without merging**